tests/test_authority.py

```diff
@@ -123,3 +128,101 @@ class MockResponse(object):
+        app = msal.PublicClientApplication("id", authority="https://contoso.com/adfs")
+        known_to_microsoft_validation.assert_not_called()
+        app.get_accounts()  # This could make an instance metadata call for authority aliases
+        instance_metadata.assert_not_called()
+
+    def test_by_default_b2c_should_skip_validation_and_instance_metadata(
+            self, instance_metadata, known_to_microsoft_validation, _):
+        """Not strictly required, but when/if we already supported it, we better keep it"""
+        app = msal.PublicClientApplication(
+            "id", authority="https://login.b2clogin.com/contoso/b2c_policy")
+        known_to_microsoft_validation.assert_not_called()
+        app.get_accounts()  # This could make an instance metadata call for authority aliases
+        instance_metadata.assert_not_called()
+
+@patch("msal.authority.tenant_discovery", return_value=_mocked_tenant_discovery_response)
+@patch("msal.authority.instance_discovery")
+@patch.object(msal.PublicClientApplication, "_get_instance_metadata", return_value=[])
+class TestMsalAuthorityBehaviorsWithKnownToDeveloperAuthorities(unittest.TestCase):
+
+    def test_known_authorities_should_be_configured_only_once(self, *mocks):
+        msal.PublicClientApplication.set_known_authorities(["private.cloud"])
+        with self.assertRaises(ValueError):
+            msal.PublicClientApplication.set_known_authorities(["hacked.com"])
+
+    def test_validate_authority_boolean_should_not_coexist_with_known_authorities(self, *mocks):
+        msal.PublicClientApplication.set_known_authorities(["private.cloud"])
+        with self.assertRaises(ValueError):
+            msal.PublicClientApplication(
+                "id", authority="https://unknown.com/common", validate_authority=False)
+
+    def test_known_to_developer_authority_should_skip_validation_and_instance_metadata(
+            self, instance_metadata, known_to_microsoft_validation, _):
+        msal.PublicClientApplication.set_known_authorities(["private.cloud"])
+        app = msal.PublicClientApplication("foo", authority="https://private.cloud/foo")
+        known_to_microsoft_validation.assert_not_called()
+        app.get_accounts()  # This could make an instance metadata call for authority aliases
+        instance_metadata.assert_not_called()
+
+    def test_a_known_to_microsoft_authority_should_skip_validation_but_still_use_instance_metadata(
+            self, instance_metadata, known_to_microsoft_validation, _):
+        msal.PublicClientApplication.set_known_authorities(["private.cloud"])
+        app = msal.PublicClientApplication(
+            "id", authority="https://login.microsoftonline.com/common")
+        known_to_microsoft_validation.assert_not_called()
+        app.get_accounts()  # This could make an instance metadata call for authority aliases
+        instance_metadata.assert_called_once_with()
+
+    def test_unknown_adfs_should_trigger_validation_and_instance_metadata(
```

Comment:
And will fail with a meaningful error??

Reply:
Not necessarily.

The test case here mainly focuses on demonstrating a "known-to-Microsoft validation" should be performed. The actual outcome would depend on the real validation response from the Instance Discovery endpoint:
* If the Instance Discovery endpoint decides that the input authority is valid, MSAL will proceed, and later also perform Instance Metadata acquisition, which is also demonstrated in this test case.
* If the Instance Discovery endpoint deems the input invalid, MSAL would yield a meaningful error, which is a preexisting behavior, so we do not need to test it here.

Reply:
And now I added a comment inside the test case to call this out. Closing this conversation.